### DISCUSSION TITLE
Menu Wrapping Discussion

### DISCUSSION THREAD

**Discussion by ok-reza** [SynthstromAudible/DelugeFirmware#515] [General]
Menu wrapping allows you to enter a menu, turn _counter-clockwise_ and end up at the end of the list of a particular menu.
However, there seems to be a lot of inconsistency with this between OLED and 7SEG, in many cases they behave in opposites.

### Examples
For 7SEG, if you open the main deluge menu (shift+press select encoder) and turn counter-clockwise, you will end up at the end of the list. However, in OLED, you can't go backwards to the end of the list.

Alternatively, if you browse into the 'Resample' folder. The 7SEG can not go backwards in order to go to the most recent resample, however the OLED can.

Additionally, most parameter menu options can not be accessed backwards on OLED. On 7SEG, I can make one turn backwards to get to the end of a parameter menu, ie. the Delay Sync options, however on OLED I can't and would have to make over 25 turns to get to the end of the list.

### What's ideal?
If a menu item only has two entries (ie. OFF or ON), should you be able to scroll either direction indefinitely with the two options continuously being selected/deselected (7SEG behavior) or should you be met with the end of the list after one turn and not be able to continue turning (OLED behavior)?

> (OLED's behavior is inconsistent when changing Delay Stereo, where you can freely spin the encoder and it will warp for DISABLE/ENABLE options. It seems like in most other cases when changing parameters, wrapping doesn't occur.)

The benefit of OLED's usual behavior (not wrapping) is that you can enter a menu of two entries, say, Delay Analog, and spin the select encoder as much as you want clockwise and you'll know for a fact that you'll end up at the 2nd entry, Analog Delay. You could make this same argument for menu items with more than two entries too, if you know what's at the end of the list you can guarantee to get to that option very quickly by spinning select a ton without any precision to that last entry.

What's great about menu wrapping (most of 7SEG's behavior) is that you can navigate menu's faster. I can easily turn select back once or twice in the main menu to get to common locations I visit in the menu (checking firmware, turning off and on quantization in the 'recording' menu, and now the community features menu.) It also is nice to know that no matter which way you turn the select knob (especially for menu's with only two entries) you will successfully select another option.

### Should OLED and 7SEG be approached differently?
I think because of OLED's display, not having menu wrapping for some menus isn't completely awful since you can easily read the entries. Additionally, for OLED, menu wrapping is pretty clear in large menus because you can visually see the highlighted selection shift from the 1st entry to the 3rd, and vice versa. In 7SEG you don't have any indication that you have wrapped around the menu aside from recognizing the same text on the display. I don't know if that means OLED and 7SEG should behave differently from each other though, I think ideally it's consistent across the board, right?

I think the easy answer is have menu wrapping for everything. But I think there should be discussion of what people prefer in which situations, if there are some edge cases that should be considered. I think we could probably all agree menu wrapping should DEFINITELY be implemented for 7SEG in the 'RESAMPLE' folder since you often need to get to the last entry and currently that folder isn't wrappable in 7SEG but is for OLED.

Of course there's always the option to have a specific 'menu wrapping' entry in the menus in order to toggle wrapping to be on or off in various situations: Sample folders, Settings Menu, Parameter Menus, Menu's with 2 entries, Menus greater than 2 entries, etc. but it seems like a messy and convoluted solution.

**Comment by sapphire-arches** [comment]
There is a decent amount of code from the original open source release to make this behavioural difference happen. However, I agree that it doesn't make a whole lot of sense, and the 7seg behavior (wrapping) is generally better.

**Reply by sapphire-arches** [reply]
in general, the whole UI system could benefit from a consistency pass...

**Comment by worker-gray** [comment]
Maybe an option for 'jump' scrolling where a modifier key pushed while scrolling would jump between first characters alphabetically-
A, B, C as first character entry and so-on.
